### PR TITLE
Issue 3384 warn when draw is not use in composite

### LINKED ISSUE #3384: Warn when an `@st.composite` strategy doesn't use the `draw()` function
I've seen a few cases where users have missed the point of `@st.composite`, and ended up using it as a substitute for `st.just()`:

```python
@st.composite
def foo(_):
    return "bar"
```

I propose to fix this by using `inspect.getsource()`+`ast.parse()`+a custom AST visitor, which will raise an explanatory error (initially a deprecation warning) if the `draw` argument is not referenced in the function body.  Note that the name of said argument is determined by the AST of the function signature, and the `ctx` of the node must be `Load()`.

### PR DESCRIPTION
**Pull request by bomtall**
This will fix #3384

We had to change tests for composite that did not reference draw

### CODE REVIEW

**Comment by rodrigogiraoserrao** [comment]
@Zac-HD we are struggling to parse the output of the failing CI action (check-format) so we have no idea what the formatting issue is.

File: hypothesis-python/tests/cover/test_reflection.py
Comment (Zac-HD) [inline hypothesis-python/tests/cover/test_reflection.py:632]: Asserting on `is` is generally a little verbose for us; prefer `assert is_called(...)` or `assert not is_called(...)`.

File: hypothesis-python/src/hypothesis/internal/reflection.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/internal/reflection.py:229]: `inspect.getsource` or `ast.parse` might fail (weird but true!), so we'll want to catch an exception around that and only run the check if we got the tree.

**Review by Zac-HD — COMMENTED** [review]
Great start!  Thanks everyone 😍

File: hypothesis-python/src/hypothesis/internal/reflection.py
Comment (rodrigogiraoserrao) [reply hypothesis-python/src/hypothesis/internal/reflection.py]: Right. Fixed in #87424d2be839333d9b975f81a58452d5c79a335e.

The docs say that `ast.parse` might throw a `ValueError` if the null character shows up in the source, but I don't think that case can happen with strings that come from a successful call to `inspect.getsource`. I do not know if there are other cases in which `ast.parse` throws a `ValueError`, so I still capture that error in the `try-except` block.

File: hypothesis-python/src/hypothesis/internal/reflection.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/internal/reflection.py:240]: Still fails if an exception was raised!  It might be simpler overall to use `ast.walk` rather than a `NodeVisitor` class?

Comment (rodrigogiraoserrao) [reply hypothesis-python/src/hypothesis/internal/reflection.py]: 🤦 right, the fix was incomplete.
Either way, you are right that an `ast.walk` simplifies the logic a bit. Fixed in 8de3c0c5d2e937b8a00c9cde12b1d1e5e2c6bfa7.

File: hypothesis-python/src/hypothesis/internal/reflection.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/internal/reflection.py:228]: The simplest way to trigger this for full coverage is probably to call it on a builtin, e.g. `is_func_param_called_within(object, "")`.

File: hypothesis-python/src/hypothesis/internal/reflection.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/internal/reflection.py:233]: Can we expand this to check for a [`Call` node where the `func` attribute](https://greentreesnakes.readthedocs.io/en/latest/nodes.html#Call) matches this condition?

File: hypothesis-python/src/hypothesis/strategies/_internal/core.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/strategies/_internal/core.py:1537]: Delete stale comment.

File: hypothesis-python/src/hypothesis/strategies/_internal/core.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/strategies/_internal/core.py:1540]:
```suggestion
            "There is no reason to use @st.composite on a function which "
            "does not call the provided draw() function internally.",
            since="RELEASEDAY",
```

File: hypothesis-python/tests/cover/test_annotations.py
Comment (Zac-HD) [inline hypothesis-python/tests/cover/test_annotations.py:86]: Let's rename this, since it's no longer pointless.  Maybe `to_wrap_with_composite`?

File: hypothesis-python/tests/cover/test_composite.py
Comment (Zac-HD) [inline hypothesis-python/tests/cover/test_composite.py:132]: Let's keep the old test, and just insert a `draw(st.none())` to keep the check happy - I think this would break the regression test as-is.

File: hypothesis-python/tests/cover/test_reflection.py
Comment (Zac-HD) [inline hypothesis-python/tests/cover/test_reflection.py:635]: It's OK if this is deprecated; I'd rather check that it's called than allow reassignment.

**Review by Zac-HD — COMMENTED** [review]
This is looking really good - thanks again for contributing!

File: hypothesis-python/src/hypothesis/internal/reflection.py
Comment (rodrigogiraoserrao) [reply hypothesis-python/src/hypothesis/internal/reflection.py:233]: We thought long and hard about this one. What if someone calls `draw` in some dynamic way that isn't a direct function call?
For example, quite insane, but `draw.__call__(...)` would still be valid.
An even more extreme case would be the code `eval("draw")(...)` which doesn't even reference the name `draw` but would work....

File: hypothesis-python/src/hypothesis/internal/reflection.py
Comment (rodrigogiraoserrao) [reply hypothesis-python/src/hypothesis/internal/reflection.py:228]: We ended up calling it on a `compile`d string. Do we change it or leave it as-is?

File: hypothesis-python/src/hypothesis/internal/reflection.py
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/reflection.py:233]: Then they will find that their insane calling convention is deprecated, and they have to change it (or add an `if False: draw()`)!

This is a very good argument against making this an error immediately, but deprecation warnings are OK 🙂

File: hypothesis-python/src/hypothesis/internal/reflection.py
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/reflection.py:228]: Compiled string is great too 👍

File: hypothesis-python/src/hypothesis/internal/reflection.py
Comment (rodrigogiraoserrao) [reply hypothesis-python/src/hypothesis/internal/reflection.py:233]: My two examples are weird, but show that there are ways of calling `draw` indirectly.
You are the maintainer, so I am assuming directly checking for a function call is sensible.

Comment (rodrigogiraoserrao) [reply hypothesis-python/src/hypothesis/internal/reflection.py:233]: I changed the check to look for an explicit function call. It works great but now breaks the [test `test_can_use_pure_args` from `tests/cover/test_composite.py`](https://github.com/HypothesisWorks/hypothesis/blob/eed37f681b2baf007034a453193d8e314c3c15f4/hypothesis-python/tests/cover/test_composite.py#L78):

https://github.com/HypothesisWorks/hypothesis/blob/eed37f681b2baf007034a453193d8e314c3c15f4/hypothesis-python/tests/cover/test_composite.py#L78-L83

I'm assuming that “using pure args” is now deprecated behaviour and should be removed from the tests?

File: hypothesis-python/src/hypothesis/internal/reflection.py
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/reflection.py]: I'd just `except Exception:`, because we might also get a `SyntaxError` etc. (e.g. from Cython code)

File: hypothesis-python/src/hypothesis/internal/reflection.py
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/reflection.py:233]: Um.  Let's skip the new check in the case that `params = get_signature(f).parameters`; `if params and list(params)[0].kind is inspect.Parameter.VAR_POSITIONAL: <skip this check>`

Messy, but I guess we're committed to this working 😅

Comment (rodrigogiraoserrao) [reply hypothesis-python/src/hypothesis/internal/reflection.py]: Understood 👌

**Comment by rodrigogiraoserrao** [comment]
I would like to point out that, as it stands, this composite is illegal:

```py
@st.composite
def allow_many_strats(draw, *strats):
    return tuple(map(draw, strats))
```

But this one is legal:

```py  
@st.composite
def allow_many_strats_ugly(*everything):
    draw, *strats = everything
    return tuple(map(draw, strats))
```

Much worse, this one is legal:

```py
@st.composite
def allow_many_strats_terrible(*everything):
    return tuple(map(everything[0], everything[1:]))
```

Not to mention using `everything[0](...)` to access “`draw`” and calling it immediately after!
For what is worth, I think allowing composite to be applied to something whose signature starts with `def foo(*args, )` is really weird! Of course, you might have a specific use case or two that I do not know of.

**Comment by Zac-HD** [comment]
😭😭😭 

I think this is actually a good argument for just checking for Name, not Call.  I'm sorry for leading you on like this, but let's go back to your previous solution!

**Comment by rodrigogiraoserrao** [comment]
Just to make sure, this now allows for

```py
@st.composite
def allow_many_strats(draw, *strats):
    return tuple(map(draw, strats))
```

but the other two remain legal. This is ok, correct?

**Comment by rodrigogiraoserrao** [comment]
Just to make sure, this now allows for

```py
@st.composite
def allow_many_strats(draw, *strats):
    return tuple(map(draw, strats))
```

but the other two remain legal. This is ok, correct?

Also, the commit history is such a mess 😢  How are you going to handle this?

**Review by Zac-HD — APPROVED** [review]
Thanks so much for your work today @gzaxel @rodrigogiraoserrao @bomtall!  This is a great PR and I'm looking forward to shipping it later tonight 😁

**Comment by Zac-HD** [comment]
> Also, the commit history is such a mess 😢 How are you going to handle this?

...squash-merge?  I tried interactive rebasing briefly but it didn't quite preserve attribution properly.

**Comment by rodrigogiraoserrao** [comment]
> > Also, the commit history is such a mess 😢 How are you going to handle this?
> 
> ...squash-merge? I tried interactive rebasing briefly but it didn't quite preserve attribution properly.

I think two of us will not be attributed if you squash-merge, which really is the more sensible thing to do. Could you hold off on that for 24h, to see if we want (and manage to) make a **much** cleaner PR out of another branch?

**Comment by Zac-HD** [comment]
Sure, happy to hold off for a while.  Thanks for your continued care on this!

**Comment by rodrigogiraoserrao** [comment]
Superseded by #3412. This can be closed 👍

**Pull request closed without merging**
